Frogger game by Ivan Chukhalo (the task is a part of #9 'Object Oriented JS' module)

ivan-chukhalo opened the pull request:
### Frogger game

[Demo](https://ivan-chukhalo.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/ivan-chukhalo/frontend-nanodegree-arcade-game/tree/develop)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

ivan-chukhalo commented:
@OleksiyRudenko 

I've run through my work, made few changes. Seems it does not have common mistakes now.
Please, review.

OleksiyRudenko commented on submissions/ivan-chukhalo/frogger-game/app.js:
```diff
@@ -0,0 +1,116 @@
+const AVAILABLE_ROWS_FOR_ENEMIES = [63, 146, 229]; 
+const allEnemies = [];
+
+const WIDTH_OF_GAME_SPRITES = 100;
+const COLUMNS_ON_GAME_FIELD = 5;
+const HEIGH_OF_GAME_FIELD_CELL = 83;
+const PLAYER_BUG_CENTERS_DISTANCE_OF_TOUCHING = 80;
+const Y_OF_GAME_FIELD_ROWS = [63, 146, 229, 312, 395];
```
There must be some formula to calculate pixel Y from row number...
Good news is that computers are made to do calculations for us. We need to benefit from this fact to make our lives better.

OleksiyRudenko reviewed (approved):
@ivan-chukhalo great job!

OleksiyRudenko commented on submissions/ivan-chukhalo/frogger-game/app.js:
```diff
@@ -0,0 +1,116 @@
+};
+
+Enemy.prototype.doesHitPlayer = function (){
+    if (Math.abs(this.x - player.x) <= PLAYER_BUG_CENTERS_DISTANCE_OF_TOUCHING && this.y === player.y){   
```
>     * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
      (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
      Hint: pass `Player` instance as an argument to every enemy

OleksiyRudenko reviewed (changes requested):
@ivan-chukhalo an improtant fix to do.

ivan-chukhalo commented:
Thank you for comments. I've made requested changes.

OleksiyRudenko commented on submissions/ivan-chukhalo/frogger-game/app.js:
```diff
@@ -0,0 +1,121 @@
+
+Enemy.prototype.update = function(dt) {     
+    this.x = this.x >= this.xAxisMax ? this.x = this.xInitial : this.x + dt * this.speedKoef;
+    this.doesHitPlayer(player);   
```
`player` is still a global variable.
The requirements provide a hint on how to resolve the problem.

ivan-chukhalo replied:
@OleksiyRudenko, excuse me.

It seems like I've missed this row or I got rid of global variables dependencies in the project, then made some extra changes in the code and added new dependency without noticing it)

Fixed issue. Review, please.

OleksiyRudenko replied:
This line of code seems to remain unchanged. `player` is a global variable.

OleksiyRudenko reviewed (changes requested):
@ivan-chukhalo nearly there

ivan-chukhalo commented:
@OleksiyRudenko 

**User Acceptance Tests** have been conducted and no bugs have been found.

OleksiyRudenko commented:
Note that UAT doesn't make issues pointed out by a mentor obsolete.

ivan-chukhalo commented:
@OleksiyRudenko 

Added an income parameter (refered to another class entity) for a class' function and forgot to rewrite function's body in previous commit😱😩. Excuse me.

I fixed the problem (as I consider), ran through code and made few extra changes to fix things which I thought could be mistakes. Basically, I did an extra self check.

OleksiyRudenko commented on submissions/ivan-chukhalo/frogger-game/app.js:
```diff
@@ -0,0 +1,112 @@
+    
+let player = new Player();
+for (let elem of AVAILABLE_ROWS_FOR_ENEMIES.concat(AVAILABLE_ROWS_FOR_ENEMIES)){
+    allEnemies.push(new Enemy(elem, player));
```
`elem` is a bad naming

OleksiyRudenko reviewed (approved):
@ivan-chukhalo great job!
